Reachability analysis: optimize resolution of reflective proxies.

sjrd opened the pull request:
There are several major changes, contributing together to a 20x speedup (!) of the reachability analysis time on our test suite. The speedup may not be representative of more standard codebases, which we expect not to have so many reflective calls in the first place.

---

1. Compute ancestors for reflective lookup once per class.

We extract the computation of all the ancestors, in the correct lookup order, in a private `lazy val` of type `List[ClassInfo]`.

---

2. Compute all the proxy target candidates for a class at once.

Iterating over all the public methods every time we need the candidates for a proxy is expensive. In the worst case, it can lead to a O(NxM) factor where N is the number of reflective calls, and M the number of methods in a class.

We now build a map of all the possible candidates for a class the first time one proxy is requested. Building the map is in theory O(M), and then each lookup becomes O(1). This turns the overall worst-case complexity from O(NxM) to O(N+M).

---

3. Fast, synchronous path for 0 and 1 proxy candidates.

When there is 0 or 1 proxy candidates, we can avoid all the complicated machinery to compute the most specific one. This is significant because that machinery involves Future computations.

We even push the call to `workQueue.enqueue` in the case with more than one candidate. This avoids most uses of `workQueue.enqueue` for reflective proxy reasons.

---

These optimizations were guided by a flame graph analysis.

gzm0 reviewed (commented):
Have you considered merging the two first commits and avoid the async loop altogether if we find a candidate where we do not need specificity evaluation?

sjrd commented:
Actually I have, and although it made this method disappear from the flame graph, it *increased* the overall clock time back to what it is after the first commit. I pushed it here if you want to have a look:
https://github.com/sjrd/scala-js/tree/analyzer-proxy-lookup-no-async-loop

I suspect that the async loop discharges the "main thread", allowing other stuff to happen in the meantime. But that explanation is probably bogus, because the async loop calls back into futures that must happen on the main thread, so :man_shrugging:

sjrd commented:
I managed to make the loop synchronous after several other changes. See #4813. The changes there build on top of this PR anyway.

sjrd commented:
Squashed everything, except the refactoring commit, because it provides a better diff, IMO.

Regarding comments at #4813:

> It seems we have a trade-off here in how lazy we should be: We're now always calculating the whole ancestor chain, even if we don't use it. IIUC this depends on how "reflective proxy heavy" a codebase is.
> 
> Have you considered using a `Stream` / `LazyList`? IIUC if we get comparable performance on the Scala.js test suite, we're pretty sure it'll be at least as good on less reflective call heavy codebases.

If we compute `ancestorsInReflectiveTargetOrder` for a given class, we also compute `proxyCandidates` for that class (the converse is not always true). Computing the latter should dwarf the computation of the former, so I don't think being more lazy here can win anything.

> With this refactoring, is the unrolling of the loop done in https://github.com/scala-js/scala-js/commit/195090eced050090990713d013942a4580d97ebf still justified?

That's now squashed away.

> At this point, you don't need the handrolled loop anymore, right? (this is just a "get the first Some in an iterator").
> 
> Does it make sense to drop the first commit of #4811?

Sort of squashed away, although there's still a handrolled loop as `findFirstNonEmptyCandidates`. Reverting that one to
```scala
      val candidates = ancestorsInReflectiveTargetOrder
        .iterator
        .map(_.findProxyCandidates(proxyName))
        .find(!_.isEmpty)
        .getOrElse(Nil)
```
causes a 10-35% slowdown, though. So I think we should keep `findFirstNonEmptyCandidates` as it is.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:
```diff
@@ -765,32 +765,79 @@ private final class Analyzer(config: CommonPhaseConfig,
+    private def findProxyCandidates(proxyName: MethodName): List[MethodInfo] =
+      proxyCandidates.getOrElse(proxyName, Nil)
+
+    private lazy val proxyCandidates = {
```
This body is so ugly, but I guess it's faster :(

gzm0 reviewed (approved):
Nice speedup. A pity we need to sacrifice conciseness: shrug:
